### DISCUSSION TITLE
Cloud control for IR devices

### DISCUSSION THREAD

**Discussion by Ameb** [jasonacox/tinytuya#283] [Q&A] [answered]
Hello
I see controlling smart IR devices is possible locally (#74)
Would it be possible to use the cloud to send `send_ir` commands? How would you translate 
```
command = {
    "control": "send_ir",
    "head": "",
    "key1": "[[TO_BE_REPLACED]]",
    "type": 0,
    "delay": 300,
}
```
To use the Cloud [`sendcommand` function](https://github.com/jasonacox/tinytuya/blob/0b7a87ad0bec813862a5843ce38a7957b8934089/tinytuya/Cloud.py#L428)

Discussion #104 is similar but aimed at subdevices created when adding a device to the Smart IR. I don't think it is possible to send the command to the subdevice via cloud.

**Comment by uzlonewolf** [comment, marked as answer]
```python
import tinytuya
import json

tinytuya.set_debug()

c = tinytuya.Cloud()

device_id = '...'

ir_cmd = {
    "control":"send_ir",
    "head":"010ed20000000000040015004000ad0730",
    "key1":"002$$0020E0E0E01F@%",
    "type":0,
    "delay":300
}

cloud_cmd = {
  "commands": [
    {
      "code": "ir_send",
      "value": json.dumps(ir_cmd)
    },
  ]
}

res = c.sendcommand(device_id, cloud_cmd)

print(res)
print('')
print( json.dumps(res, indent=2) )
```

If you have remotes programmed/defined you can also send keys:

```python
import tinytuya
import json

tinytuya.set_debug()

c = tinytuya.Cloud()

device_id = '...' # Smart IR (IR blaster)

# Get a listing of all programmed remotes
#res = c.cloudrequest( '/v2.0/infrareds/' + device_id + '/remotes' )

# Get a list of all keys on a remote
remote_id = '...'
#res = c.cloudrequest( '/v2.0/infrareds/' + device_id + '/remotes/' + remote_id + '/keys' )

# Send a key
post_data = {
  "category_id": 2,
  "remote_index": 10572,
  "key": "Power"
}
#res = c.cloudrequest( '/v2.0/infrareds/' + device_id + '/remotes/' + remote_id + '/command', post=post_data )

print( res )
print( json.dumps(res, indent=2) )
```

**Reply by Ameb** [reply]
Thank you for the suggestion!
I wasn't aware of the `cloudrequest` function. It was very useful as I could try other tuya APIs like [infrareds/air-conditioners](https://developer.tuya.com/en/docs/cloud/infrared-air-conditioner-apis?id=Kb3oe9ehg02fn)
```python
payload={
    "remote_index": 101352128
    ,"category_id": 5
    ,"power": 1
    ,"mode": 1
    ,"temp": 30
    ,"wind": 2
}
res = c.cloudrequest( '/v2.0/infrareds/' + device_id + '/air-conditioners/' + remote_id + '/scenes/command', post=payload)
```

Sadly, I'm having issues with building the `ir_cmd` part for the sendcommand. On the device logs i can see the base64 string sent to the device to send a custom DIY command:
```json
{
"control":"send_ir",
"head":"",
"key1":"10CRwEXMCPgI+Aj4CcwI+AnMCPgJzAj4CcwI+Aj4CPgI+ApEGPgKRBj4CkQZzApEGcwKRBj4CkQY+ApEGPgKRBnMCPgJzApEGcwI+AnMCPgI+ApEGPgI+Aj4CPgI+Aj4CPgI+Aj4CPgI+ApEGcwKRBnMCPgI+ApEGPgKRBj4CkQY+ApEGcwKImtAk6Ag+AjB1",
"type":0,
"delay":300
}
```

I know i can transform the base64 string into hex with the static methods on [`IRRemoteControlDevice`](https://github.com/jasonacox/tinytuya/blob/master/tinytuya/Contrib/IRRemoteControlDevice.py)
```python
from tinytuya.Contrib import IRRemoteControlDevice

key1_b64="10CRwEXMCPgI+Aj4CcwI+AnMCPgJzAj4CcwI+Aj4CPgI+ApEGPgKRBj4CkQZzApEGcwKRBj4CkQY+ApEGPgKRBnMCPgJzApEGcwI+AnMCPgI+ApEGPgI+Aj4CPgI+Aj4CPgI+Aj4CPgI+ApEGcwKRBnMCPgI+ApEGPgKRBj4CkQY+ApEGcwKImtAk6Ag+AjB1"
# tried with both ...AjB1 and ...AjB (without the '1' padding)
ir_cmd = {
    "control":"send_ir",
    "head":"",
    "key1":IRRemoteControlDevice.pulses_to_hex(IRRemoteControlDevice.base64_to_pulses(key1_b64)), #tried with the base64 string too
    "type":0,
    "delay":300
}

cloud_cmd = {
  "commands": [
    {
      "code": "ir_send",
      "value": json.dumps(ir_cmd)
    },
  ]
}

res = c.sendcommand(device_id, cloud_cmd)

print(res)
print('')
print( json.dumps(res, indent=2) )
```
But I must be missing something:
```
DEBUG:tinytuya.core:POST: URL=https://openapi.tuyaeu.com/v1.0/iot-03/devices/<device_id>/commands HEADERS={'Content-type': 'application/json', 'Signature-Headers': 'Content-type', 'client_id': '', 'sign': '', 't': '1677320500047', 'sign_method': 'HMAC-SHA256', 'access_token': ''} DATA={"commands": [{"code": "ir_send", "value": "{\"control\": \"send_ir\", \"head\": \"\", \"key1\": \"d74091c045cc08f808f808f809cc08f809cc08f809cc08f809cc08f808f808f808f80a4418f80a4418f80a4419cc0a4419cc0a4418f80a4418f80a4418f80a4419cc08f809cc0a4419cc08f809cc08f808f80a4418f808f808f808f808f808f808f808f808f808f808f80a4419cc0a4419cc08f808f80a4418f80a4418f80a4418f80a4419cc0a226b4093a020f808c1\", \"type\": 0, \"delay\": 300}"}]}
DEBUG:tinytuya.core:Error from Tuya Cloud: 'command or value not support'

{'code': 2008, 'msg': 'command or value not support', 'success': False, 't': 1677320500500, 'tid': '318d66d5b4f611edad0316dc0085482e'}

{
  "code": 2008,
  "msg": "command or value not support",
  "success": false,
  "t": 1677320500500,
  "tid": "318d66d5b4f611edad0316dc0085482e"
}
```

**Reply by uzlonewolf** [reply]
I'm not sure what exactly you are trying to send is?  When using base64 it should be passed directly in key1.  This works fine for me:

```python
import tinytuya
import json

#tinytuya.set_debug()

device_id = '...'
c = tinytuya.Cloud(...)

ir_cmd = {
    "control":"send_ir",
    "head":"",
    "key1":"10CRwEXMCPgI+Aj4CcwI+AnMCPgJzAj4CcwI+Aj4CPgI+ApEGPgKRBj4CkQZzApEGcwKRBj4CkQY+ApEGPgKRBnMCPgJzApEGcwI+AnMCPgI+ApEGPgI+Aj4CPgI+Aj4CPgI+Aj4CPgI+ApEGcwKRBnMCPgI+ApEGPgKRBj4CkQY+ApEGcwKImtAk6Ag+AjB1",
    "type":0,
    "delay":300
}

cloud_cmd = {
    "commands": [
        {
            "code": "ir_send",
            "value": json.dumps(ir_cmd)
        },
    ]
}

res = c.sendcommand(device_id, cloud_cmd)

print(res)
print( json.dumps(res, indent=2) )
print('')

print('Device logs:')
logs = c.getdevicelog(device_id, evtype='5', size=5, max_fetches=1)
print( json.dumps(logs, indent=2) )
```
Result:
```
{'result': True, 'success': True, 't': 1677336560098, 'tid': '95xxxxxxxxxxxxxxxxxxxxxxxxxe'}
{
  "result": true,
  "success": true,
  "t": 1677336560098,
  "tid": "95xxxxxxxxxxxxxxxxxxxxxxxxe"
}

Device logs:
{
  "result": {
    "current_row_key": "NDxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "device_id": "ebxxxxxxxxxxxxxxxxx",
    "has_next": false,
    "logs": [
      {
        "code": "ir_send",
        "event_from": "-1",
        "event_id": 5,
        "event_time": 1677336154776,
        "status": "1",
        "value": "{\"control\": \"send_ir\", \"head\": \"\", \"key1\": \"10CRwEXMCPgI+Aj4CcwI+AnMCPgJzAj4CcwI+Aj4CPgI+ApEGPgKRBj4CkQZzApEGcwKRBj4CkQY+ApEGPgKRBnMCPgJzApEGcwI+AnMCPgI+ApEGPgI+Aj4CPgI+Aj4CPgI+Aj4CPgI+ApEGcwKRBnMCPgI+ApEGPgKRBj4CkQY+ApEGcwKImtAk6Ag+AjB1\", \"type\": 0, \"delay\": 300}"
      }
    ]
  },
  "success": true,
  "t": 1677336560434,
  "tid": "96xxxxxxxxxxxxxxxxxe",
  "fetches": 1
}
```

The only time you need to do any decoding is if you want to turn it into a head/key1 pair.  This is completely optional and it will work just fine if you leave it as base64.  If you do want to convert it, then:

```python
from tinytuya.Contrib import IRRemoteControlDevice

b64 = '10CRwEXMCPgI+Aj4CcwI+AnMCPgJzAj4CcwI+Aj4CPgI+ApEGPgKRBj4CkQZzApEGcwKRBj4CkQY+ApEGPgKRBnMCPgJzApEGcwI+AnMCPgI+ApEGPgI+Aj4CPgI+Aj4CPgI+Aj4CPgI+ApEGcwKRBnMCPgI+ApEGPgKRBj4CkQY+ApEGcwKImtAk6Ag+AjB1'

pulses = IRRemoteControlDevice.base64_to_pulses( b64 )
#print(pulses)
headkey = IRRemoteControlDevice.pulses_to_head_key( pulses, fudge=0.2 )
#print(headkey)
print( 'head:', headkey[0] )
print( 'key:', headkey[1] )
```
```
head: 010ED800000000000700170040016600AA05DF00570474
key: 01$%002001FE906F@^$&@*
```
Seems to be a standard 32-bit NEC code.  If you are going to pass it to send_ir yourself you will need to add a "0" to the beginning of key1:
```python
ir_cmd = {
    "control":"send_ir",
    "head":"010ED800000000000700170040016600AA05DF00570474",
    "key1":"001$%002001FE906F@^$&@*",
    "type":0,
    "delay":300
}
```

**Reply by Ameb** [reply]
The code is for a simple two button KVM controller. I'm wondering why the command (with the base64 string) doesn't work on this tuya device.
```json
{
  "code": 2008,
  "msg": "command or value not support",
  "success": false,
  "t": 1677365444583,
  "tid": "d64xxxxxxxxxxxxxxxxxxxxxxxxx"
}
```

About the device:
```
Product Name
S06CBU（S18通用）红外遥控器
Product Category
wnykq
```

I'm very happy with your answer and this topic will probably help others. The `cloudrequest` will solve my initial needs using all the API endpoints available.
If I have the time I will test the `sendcommand` method with another IR tuya device I have.

**Reply by uzlonewolf** [reply]
> S06CBU

The closest I have in my collection is a HMS06CBU, and of course I don't know where it is at the moment.  Looking at the cloud configuration however it looks like `send_ir` is not exposed in the Standard Instruction instruction set.
![Screenshot_2023-02-25_20-01-37](https://user-images.githubusercontent.com/16113541/221391696-bddc836b-da7c-42cd-8426-7ae69aeffd2a.png)
This is different than my S06ProWB3S which does expose `send_ir` in Standard.

To change to the DP instruction set, follow https://github.com/jasonacox/tinytuya/discussions/284#discussioncomment-4953888 and wait ~24 hours for the change to propagate before trying again.

**Reply by uzlonewolf** [reply]
I just found it and can confirm, CBU modules reject `send_ir` commands when set to Standard Instruction set.
```
ir_cmd = {
    "control":"send_ir",
    "head":"010ed20000000000040015004000ad0730",
    "key1":"002$$0020E0E0E01F@%",
    "type":0,
    "delay":300
}

cloud_cmd = {
    "commands": [
        {
            "code": "ir_send",
            "value": json.dumps(ir_cmd)
        },
    ]
}

Send Raw result:
{
  "code": 2008,
  "msg": "command or value not support",
  "success": false,
  "t": 1677426786041,
  "tid": "a89xxxxxxxxxxxxxxxxxx2"
}
```
Both `c.getfunctions(device_id)` and `c.getproperties(device_id)` return empty results as well (my other IR blasters have results containing "function: code: ir_send").  I just changed it to the DP Instruction set and am waiting for the change to propagate; I fully expect it to start working once that goes thorough.

**Reply by Ameb** [reply]
Can confirm `getfunctions`/`getproperties` returns empty results and the device is set to "Standard Instruction"
![image](https://user-images.githubusercontent.com/386567/221436217-3b7870b5-81b6-4efd-926a-6e823ec67471.png)

I could not test my other device yet but the sendcommand will proably work as it is set in DP mode by default
![image](https://user-images.githubusercontent.com/386567/221436309-3e92baf0-be61-4eee-b17e-34c316c41f14.png)

**Reply by uzlonewolf** [reply]
My CBU module is responding to `send_ir` now that the change has gone through.  I'll update the cloud_ir.py example to note that the instruction set may need to be changed to make everything work.

**Comment by fabianoarruda** [comment]
Not sure if this is the correct place, but I made a comment on an issue and I realized it is closed already, so this discussion here may be the correct place for it. I think I found a new pattern for the `ir_send`, DPS 201. Please take a look here: https://github.com/jasonacox/tinytuya/issues/74#issuecomment-2053762869
